lib/src/authentication/credential.dart

```diff
@@ -44,14 +54,19 @@ class Credential {
 
     final hostedUrl = validateAndNormalizeHostedUrl(json['url'] as String);
 
-    const knownKeys = {'url', 'token'};
+    const knownKeys = {'url', 'token', 'env'};
     final unknownFields = Map.fromEntries(
         json.entries.where((kv) => !knownKeys.contains(kv.key)));
 
+    String _optional(String key) {
+      return json[key] is String ? json[key] as String : null;
+    }
+
     return Credential._internal(
       url: hostedUrl,
-      token: json['token'] is String ? json['token'] as String : null,
       unknownFields: unknownFields,
+      token: _optional('token'),
```

Comment:
What happens if there is both a token and an env key in the input? I guess currently we check for the env-var first. Should we instead consider that invalid?

Reply:
There's `Credential.isValid` method for validity check used to filter out invalid values when looking for token. I'm not checking validity in fromJson otherwise invalid values will be removed from pub-tokens.json when running token add or remove. That might cause pub-tokens.json generated by future version of SDK where contains entries with both token and env values that are used for some purpose to be removed.

Reply:
Also currently I'm using `env` value if provided otherwise `token` value [here](https://github.com/dart-lang/pub/pull/3115/files#diff-9d07b56c1956f7361cc0b0c4b37054c0035634e87f45dffc4cd3d5f2817bc00eR112-R123).

Reply:
So maybe `isValid` should also check that both are not set?

Reply:
> So maybe `isValid` should also check that both are not set?

Sure.